collect/stressRelief/stress_relief_redis.go

```diff
@@ -0,0 +1,365 @@
+	"bytes"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dgryski/go-wyhash"
+	"github.com/gofrs/uuid/v5"
+	"github.com/honeycombio/refinery/config"
+	"github.com/honeycombio/refinery/internal/gossip"
+	"github.com/honeycombio/refinery/logger"
+	"github.com/honeycombio/refinery/metrics"
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/sync/errgroup"
+)
+
+var _ StressReliever = &StressRelief{}
+
+type StressRelief struct {
+	RefineryMetrics metrics.Metrics `inject:"genericMetrics"`
+	Logger          logger.Logger   `inject:""`
+	Gossip          gossip.Gossiper `inject:"gossip"`
+	Clock           clockwork.Clock `inject:""`
+	done            chan struct{}
+
+	mode               StressReliefMode
+	activateLevel      uint
+	deactivateLevel    uint
+	overallStressLevel uint
+	sampleRate         uint64
+	upperBound         uint64
+	reason             string
+	formula            string
+	stressed           bool
+	stayOnUntil        time.Time
+	minDuration        time.Duration
+	identification     string
+
+	eg *errgroup.Group
+
+	algorithms map[string]func(string, string) float64
+	calcs      []StressReliefCalculation
+
+	lock         sync.RWMutex
+	stressLevels map[string]stressReport
+}
+
+func (s *StressRelief) Start() error {
+	s.Logger.Debug().Logf("Starting StressRelief system")
+	defer func() { s.Logger.Debug().Logf("Finished starting StressRelief system") }()
+
+	// We use an algorithms map so that we can name these algorithms, which makes it easier for several things:
+	// - change our mind about which algorithm to use
+	// - logging the algorithm actually used
+	// - making it easier to make them configurable
+	// At the moment, we are not permitting these to be configurable, but we might change our minds on this.
+	// Thus, we're also including a couple of algorithms we don't currently use for convenience.
+	algorithms := stressReliefAlgorithm{
+		data:   s.RefineryMetrics,
+		logger: s.Logger,
+	}
+	s.algorithms = map[string]func(string, string) float64{
+		"linear":  algorithms.linear,  // just use the ratio
+		"sqrt":    algorithms.sqrt,    // small values are inflated
+		"square":  algorithms.square,  // big values are deflated
+		"sigmoid": algorithms.sigmoid, // don't worry about small stuff, but if we cross the midline, start worrying quickly
+	}
+
+	// All of the numerator metrics are gauges. The denominator metrics are constants.
+	s.calcs = []StressReliefCalculation{
+		{Numerator: "collector_incoming_queue_length", Denominator: "INCOMING_CAP", Algorithm: "sqrt", Reason: "CacheCapacity (incoming)"},
+		{Numerator: "libhoney_upstream_queue_length", Denominator: "UPSTREAM_BUFFER_SIZE", Algorithm: "sqrt", Reason: "UpstreamBufferSize"},
+		{Numerator: "memory_heap_allocation", Denominator: "MEMORY_MAX_ALLOC", Algorithm: "sigmoid", Reason: "MaxAlloc"},
+		// TODO: what's the denominator for this one?
+		// users need to tell us what's their redis memory limit
+		//{Numerator: "redisstore_memory_used_total", }
```

Comment:
Let's add a configuration value for this; maybe just add a TODO since we'll need to come back to it anyway.

Reply:
i added a TODO here